spark/src/main/java/org/opensearch/sql/spark/cluster/FlintIndexRetention.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.spark.cluster;
+
+import static org.opensearch.sql.spark.execution.session.SessionModel.LAST_UPDATE_TIME;
+import static org.opensearch.sql.spark.execution.statement.StatementModel.SUBMIT_TIME;
+
+import java.time.Clock;
+import java.time.Duration;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.CheckedConsumer;
+import org.opensearch.common.time.FormatNames;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.index.IndexNotFoundException;
+import org.opensearch.index.query.QueryBuilders;
+
+public class FlintIndexRetention implements Runnable {
+  private static final Logger LOG = LogManager.getLogger(FlintIndexRetention.class);
+
+  static final String SESSION_INDEX_NOT_EXIST_MSG = "Checkpoint index does not exist.";
+
+  static final String RESULT_INDEX_NOT_EXIST_MSG = "Result index does not exist.";
+
+  // timestamp field in result index
+  static final String UPDATE_TIME_FIELD = "updateTime";
+
+  private final Duration defaultSessionTtl;
+  private final Duration defaultResultTtl;
+  private final Clock clock;
+  private final IndexCleanup indexCleanup;
+  private final String sessionIndexNameRegex;
+  private final String resultIndexNameRegex;
+
+  public FlintIndexRetention(
+      Duration defaultSessionTtl,
+      Duration defaultResultTtl,
+      Clock clock,
+      IndexCleanup indexCleanup,
+      String sessionIndexNameRegex,
+      String resultIndexNameRegex) {
+    this.defaultSessionTtl = defaultSessionTtl;
+    this.defaultResultTtl = defaultResultTtl;
+    this.clock = clock;
+    this.indexCleanup = indexCleanup;
+    this.sessionIndexNameRegex = sessionIndexNameRegex;
+    this.resultIndexNameRegex = resultIndexNameRegex;
+  }
+
+  @Override
+  public void run() {
+    purgeSessionIndex();
+  }
+
+  private void purgeSessionIndex() {
+    purgeIndex(
+        sessionIndexNameRegex,
+        defaultSessionTtl,
+        LAST_UPDATE_TIME,
+        this::handleSessionPurgeResponse,
+        this::handleSessionPurgeError);
+  }
+
+  private void handleSessionPurgeResponse(Long response) {
```

Comment:
right. I do it in sequence since delete by query is not a cheap query and our purging is not time sensitive. I want to achieve the purging without too much performance impact.